src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -2,46 +2,204 @@
 
 import java.util.List;
 
-public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+public class MyLinkedList<M> implements MyLinkedListInterface<M> {
+    private Node<M> top;
+    private Node<M> bottom;
+    private int size;
+
     @Override
-    public void add(T value) {
+    public void add(M value) {
+        if (size == 0) {
+            linkTop(value);
+        } else {
+            linkBottom(value);
+        }
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(M value, int index) {
+        checkPositionIndex(index);
+        if (index == 0 || size == 0) {
+            linkTop(value);
+        } else if (index == size) {
+            linkBottom(value);
+        } else {
+            Node<M> a = node(index);
+            Node<M> newNode = new Node<>(a.prev, value, a);
+            a.prev.next = newNode;
+            a.prev = newNode;
+            size++;
+        }
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void addAll(List<M> list) {
+        if (!(list == null)) {
+            for (M item : list) {
+                add(item);
+            }
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public M get(int index) {
+        checkItemIndex(index);
+        return node(index).item;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public M set(M value, int index) {
+        checkItemIndex(index);
+        Node<M> a;
+        if (index == 0) {
+            a = top;
+            unLinkTop(a);
+            linkTop(value);
+        } else {
+            a = node(index);
+            unLink(a);
+            add(value, index);
+        }
+        return a.item;
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public M remove(int index) {
+        checkItemIndex(index);
+        if (index == 0) {
+            Node<M> b = top;
+            unLinkTop(b);
+            return b.item;
+        } else if (index == size - 1) {
+            Node<M> c = bottom;
+            unLinkBottom(c);
+            return c.item;
+        } else {
+            Node<M> a = node(index);
+            unLink(a);
+            return a.item;
+        }
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(M object) {
+        Node<M> b = top;
+        while (b != null) {
+            if (object == null && b.item == null) {
+                unLink(b);
+                return true;
+            } else if (object != null && object.equals(b.item)) {
+                if (size == 1 || object.equals(top.item)) {
+                    unLinkTop(b);
+                } else {
+                    unLink(b);
+                }
+                return true;
+            }
+            b = b.next;
```

Comment:
this method is hard to read. try to refactor it, so it should find node and pass it to unlink method